File: pixijs/game/enemy-factory.js
import { ChaserEnemy, ZigzagChaserEnemy, ShooterEnemy, WallShooterEnemy, BullEnemy, BuldygaEnemy, BloatedEnemy, CocoonEnemy, PhaseBoss, GhostEnemy } from './enemy-types.js';

export class EnemyFactory {
  static create(type, x, y, options = {}) {
    const data = {
      type,
      x,
      y,
      hp: options.hp || EnemyFactory.getDefaultHp(type, options.level),
      radius: options.radius || EnemyFactory.getDefaultRadius(type, options.level),
      visualScale: options.visualScale || EnemyFactory.getDefaultVisualScale(type, options.level),
      ...options
    };

    switch (type) {
      case 'boss_phase':
        return new PhaseBoss(data);
      case 'shooter':
        return new ShooterEnemy(data);
      case 'wallshooter':
        return new WallShooterEnemy(data);
      case 'bull':
        return new BullEnemy(data);
      case 'buldyga':
        return new BuldygaEnemy(data);
      case 'bloated':
        return new BloatedEnemy(data);
      case 'cocoon':
        return new CocoonEnemy(data);
      case 'tank':
      case 'soldier':
        return new ChaserEnemy(data);
      case 'bat':
        return new ZigzagChaserEnemy(data);
      case 'ghost':
        return new GhostEnemy(data);
      default:
        return new ChaserEnemy(data);
    }
  }

  static fromObject(obj) {
    if (!obj || !obj.type) return null;
    return EnemyFactory.create(obj.type, obj.x, obj.y, obj);
  }

  static getDefaultHp(type, level = 1) {
    if (typeof CONFIG === 'undefined') return 10;
    if (type === 'boss_phase' && typeof BOSS_DEFS !== 'undefined') {
      const def = BOSS_DEFS[level] || BOSS_DEFS[1];
      if (def.hp !== undefined) return def.hp;
      const base = def.hpBase === 'buldyga' ? CONFIG.ENEMY_STATS.buldyga.hp : CONFIG.ENEMY_STATS.soldier.hp;
      return base * (def.hpMult || 1);
    }
    const stats = CONFIG.ENEMY_STATS[type];
    if (stats) return stats.hp;
    return CONFIG.ENEMY_STATS.soldier.hp;
  }

  static getDefaultRadius(type, level = 1) {
    if (typeof CONFIG === 'undefined') return 20;
    if (type === 'boss_phase' && typeof BOSS_DEFS !== 'undefined') {
      const def = BOSS_DEFS[level] || BOSS_DEFS[1];
      return (def.radius || CONFIG.ENEMY_STATS.soldier.radius) * (def.radiusMult || 1);
    }
    const stats = CONFIG.ENEMY_STATS[type];
    if (stats) return stats.radius;
    return CONFIG.ENEMY_STATS.soldier.radius;
  }

  static getDefaultVisualScale(type, level = 1) {
    if (typeof CONFIG === 'undefined') return 3.2;
    if (type === 'boss_phase' && typeof BOSS_DEFS !== 'undefined') {
      const def = BOSS_DEFS[level] || BOSS_DEFS[1];
      return def.visualScale || 4.0;
    }
    const stats = CONFIG.ENEMY_STATS[type];
    if (stats) return stats.visualScale || 3.2;
    return CONFIG.ENEMY_STATS.soldier.visualScale || 3.2;
  }
}

File: Scripts/Actors/EnemyFactory.cs
using Godot;
using SpaceOrLife.Core.Catalogs;
using SpaceOrLife.Domain;
using SpaceOrLife.World;

namespace SpaceOrLife.Actors;

public static class EnemyFactory
{
    public static EnemyActor Create(
        EnemyKind kind,
        Vector2 pos,
        EnemyCatalog catalog,
        int level,
        int roomIdx,
        bool isBoss = false)
    {
        var def = catalog.GetByKind(kind);
        var hp = (def?.Hp ?? 120f) * GetHpMult(level);
        var radius = def?.Radius ?? 7f;
        var visualScale = def?.VisualScale ?? 2.9f;

        var behavior = CreateBehavior(kind);
        if (def != null)
            behavior.SetDefinition(def);

        var actor = new EnemyActor();
        actor.Setup(kind, hp, radius, visualScale, level, roomIdx, isBoss, behavior);
        actor.GlobalPosition = pos;
        return actor;
    }

    public static EnemyActor CreateBoss(
        BossDefinition bossDef,
        Vector2 pos,
        EnemyCatalog catalog,
        int level)
    {
        var bossBehavior = new BossBehavior();
        var soldierDef = catalog.GetByKind(EnemyKind.Soldier);
        var soldierRadius = soldierDef?.Radius ?? 7f;

        var hp = BattleSystem.CalculateBossHp(bossDef, catalog);
        var radius = soldierRadius * (bossDef.RadiusMult > 0f ? bossDef.RadiusMult : 2.25f);
        var visualScale = bossDef.VisualScale > 0f ? bossDef.VisualScale : 4.0f;

        var actor = new EnemyActor();
        actor.Setup(EnemyKind.BossPhase, hp, radius, visualScale, level, -1, true, bossBehavior);
        actor.SetBossDefinition(bossDef);
        actor.GlobalPosition = pos;
        return actor;
    }

    public static EnemyActor CreateFromStasis(
        StasisEnemyData data,
        EnemyCatalog catalog)
    {
        var actor = Create(data.Type, new Vector2(data.X, data.Y), catalog, data.Level, data.RoomIdx);
        actor.Stasis = true;
        return actor;
    }

    private static EnemyBehavior CreateBehavior(EnemyKind kind)
    {
        return kind switch
        {
            EnemyKind.Soldier => new ChaserBehavior(),
            EnemyKind.Tank => new ChaserBehavior(),
            EnemyKind.Bat => new ZigzagChaserBehavior(),
            EnemyKind.Shooter => new ShooterBehavior(),
            EnemyKind.Wallshooter => new WallShooterBehavior(),
            EnemyKind.Bull => new BullBehavior(),
            EnemyKind.Buldyga => new BuldygaBehavior(),
            EnemyKind.Bloated => new BloatedBehavior(),
            EnemyKind.Cocoon => new CocoonBehavior(),
            EnemyKind.Ghost => new GhostBehavior(),
            _ => new ChaserBehavior(),
        };
    }

    private static float GetHpMult(int level) => level switch
    {
        1 => 1f,
        2 => 2f,
        3 => 4f,
        _ => 1f,
    };
}
